Renderer - FNA

flyingpie opened the pull request:
Related to #60 

- The sample still requires a couple of dependencies to be downloaded. I prefer to not have binaries in source control, but like the cimgui one, FNA.dll isn't available through NuGet and neither are the FNA dependencies. What's your stance on the inclusion of said binaries? (FNA.dll is included, but its dependencies are not)

- If you're up for it, I'll add a NuGet spec as well, but I'd like to know your preference on the naming of things first.

Whenever you have the time, feel free to take a look and let me know what you think.

mellinoe commented:
* I'd like to avoid checking in any more binary dependencies if at all possible. Is there any way we can reference some well-known location for FNA? The existing binary assets in here are at least bundled in the nuget package, so they have a better reason for being checked in.
* I'm not sure I want to have a separate ImGui.NET.FNA library / NuGet package, since it will be difficult for me to maintain and support. I don't use FNA myself so I'm not very familiar with it, and probably can't work on it very efficiently. My initial thought was that we'd just have a second sample project, and the renderer would live inside of that project. Folks could copy-paste it out to consume it, tweaking it a bit to fit their engine better.

mellinoe commented on src/ImGui.NET.FNA/ImVector.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace ImGuiNET.FNA
+{
+    public unsafe struct ImVector<T> where T : struct
```
I don't think this structure is necessary, personally. I'm planning to improve the native ImVector struct with the auto-gen rewrite, so it will be usable in much the same way.  The way this is currently being used is pretty inefficient, anyways (reading/writing individual elements).

mellinoe commented on src/ImGui.NET.FNA/ImGuiRenderer.cs:
```diff
@@ -0,0 +1,313 @@
+                var cmdList = drawData->CmdLists[n];
+
+                var vtxBuffer = new ImVector<DrawVertXNA>(&cmdList->VtxBuffer);
+                var vtxArray = new DrawVertXNA[vtxBuffer.Size];
```
This is going to produce a very large amount of garbage every frame (depending on how complex the UI is)

mellinoe replied:
To fix, you should have a persistent array that gets resized when needed.

mellinoe commented on src/ImGui.NET.FNA/ImGuiRenderer.cs:
```diff
@@ -0,0 +1,313 @@
+                var vtxArray = new DrawVertXNA[vtxBuffer.Size];
+                for (int i = 0; i < vtxBuffer.Size; i++)
+                {
+                    vtxArray[i] = vtxBuffer[i];
```
Copying over elements one-by-one is going to be very slow. Instead, you should do a single memory copy.

Ideally, this isn't necessary at all and you just use the pointer directly, but I believe that FNA doesn't support uploading vertex data from a pointer. IMO, it should, because this is still going to be a bit slower and require you to manage this separate array.

mellinoe commented on src/ImGui.NET.FNA/ImGuiRenderer.cs:
```diff
@@ -0,0 +1,313 @@
+                    if (pcmd->UserCallback != IntPtr.Zero) throw new NotSupportedException("The user callback of the command buffer is not supported");
+
+                    // Instead of uploading the complete idxBuffer again and again, just upload what's required
+                    var idxArray = new short[pcmd->ElemCount];
```
Again, this will produce lots of garbage every frame. Fixing this one will be a bit trickier because there can be lots of separate draw submissions.

mellinoe commented on src/ImGui.NET.FNA/DrawVertXNA.cs:
```diff
@@ -0,0 +1,30 @@
+namespace ImGuiNET.FNA
+{
+    [StructLayout(LayoutKind.Sequential)]
+    public unsafe struct DrawVertXNA
```
I don't think this type is really necessary. It's structurally identical to the native vertex type, and the static stuff could just be helper functions in the renderer.

mellinoe commented on src/ImGui.NET.FNA/ImGuiRenderer.cs:
```diff
@@ -0,0 +1,313 @@
+            var tex2d = texture as Texture2D;
+            if (tex2d == null) throw new InvalidOperationException($"Only textures of type '{nameof(Texture2D)}' are supported");
+
+            // The creation of a new texture id and adding it to the local dictionary should be an atomic operation, so we use a lock to ensure that (the concurrent dictionary alone isn't enough)
```
Hmm, not saying it hurts to be safe, but most everything else in ImGui is not thread-safe. It's probably unnecessary to worry about it in this particular spot.

FlyingPie commented:
@mellinoe Thank you for the quick response! I'll work through your points, and let you know when these should be fixed. Gonna do some profiling do take a look at the memory pressure.

FlyingPie commented:
@mellinoe Updated the PR:

- Moved the sample code and the renderer into 1 project
- Removed ImVector<T> and DrawVertXNA, moved vertex metadata to a static helper class
- Overhauled the rendering code, removed any copies, it now looks very much like the renderer for Veldrid
- Removed thread-safety from the texture binding stuff

The effects on the GC are profound.

Before:
![imguirenderer_before](https://user-images.githubusercontent.com/1295673/43688806-42263364-98f0-11e8-87d9-c491055e030a.png)

After:
![imguirenderer_after](https://user-images.githubusercontent.com/1295673/43688814-618f4f06-98f0-11e8-8470-8718911d03c2.png)

Looking forward to your comments.

mellinoe commented:
@FlyingPie Great, thanks for updating this! I'll try to take a look today and provide more feedback. In the meantime, it looks like your commits are mixed between two authors (and/or email addresses). I don't particularly care, but you might want to fix it if that's important to you.

mellinoe commented on src/ImGui.NET.SampleProgram.FNA/ImGuiRenderer.cs:
```diff
@@ -0,0 +1,346 @@
+                        _graphicsDevice.DrawIndexedPrimitives(
+                            primitiveType: PrimitiveType.TriangleList,
+                            baseVertex: vtxOffset,
+                            minVertexIndex: vtxOffset,
```
I don't really know what this "minVertexIndex" parameter does. Is this right? It feels weird that it's the same as "baseVertex".

FlyingPie replied:
Apparently the "baseVertex" is the actual offset in the vertex array, from which all other work is done. "minVertexIndex" can then be provided as an additional offset, relative to "baseVertex". I haven't found out why though.

The render seems to still work, even when setting the "minVertexIndex" to 0 or 100000. FNA's source describes the parameter as a "hint", and the OpenGL doc of the underlying function mentions a mechanism in which overflowing vertices are still drawn.

I've changed the minVertexIndex to 0, which is the most "correct" value I can think of. I imagine that some of these parameters may be lost in translation a bit, considering they came from a Direct3D backend.

mellinoe replied:
Interesting. I tried to look them up as well and, to the best that I could find, they are just "hints" for some legacy renderer or something like that. It seems like neither of them really matter and aren't relevant for modern API's.

flyingpie commented:
@mellinoe I wasn't happy with the image in source control, but I'd like to have an example on how to integrate FNA's textures with ImGui. So I replaced the PNG with runtime-generated image. Only code in the PR now.

mellinoe commented:
Thanks for all of the work on this. I pulled it down and didn't have any trouble figuring out how to reference FNA and the native deps. Everything seems to be working correctly as far as I can tell, and the code looks good at this point.

Instead of listing this in the README file:

> - Copy cimgui.dll from "deps" to the project output directory

We should be able to just copy these lines over from the other sample project:

```XML
  <ItemGroup>
    <Content Include="$(RepositoryRootDirectory)/deps/cimgui/win-x64/cimgui.dll" CopyToOutputDirectory="PreserveNewest" />
    <Content Include="$(RepositoryRootDirectory)/deps/cimgui/osx-x64/cimgui.dylib" CopyToOutputDirectory="PreserveNewest" />
    <Content Include="$(RepositoryRootDirectory)/deps/cimgui/linux-x64/cimgui.so" CopyToOutputDirectory="PreserveNewest" />
  </ItemGroup>
```

Then, that part should be handled automatically for you.

mellinoe commented:
Another thing I was thinking about, but which you could probably clear up. It's my understanding that this code will be mostly compatible with MonoGame as-is, since FNA and MonoGame are both re-implementations of the same API. However, there's a couple "EXT" functions used which would render it incompatible. Do you think it is worth adding comments above those functions describing what would be done in XNA/MonoGame without the EXT functions, or is that something that the community already understands how to handle?

flyingpie commented:
@mellinoe Thank you for your comments, I've cleaned up the cimgui deps, totally missed that the first time around.

Regarding MonoGame, the changes should be minimal:
- Marshal ImVector into a memory buffer
- Replace the SetDataPointerEXT() calls with SetData()

So I'll add some comments this weekend. Could you tell me the preferred way of marshaling the ImVector? Especially since that class may chance when the code generation is here.

mellinoe commented:
The best way to do it would be to have a persistent byte[] array (not recreated every frame), and just pin it and copy the data over using Buffer.MemoryCopy or a similar API.

flyingpie commented:
@mellinoe After some messing about, I changed the rendering a bit to be compatible with both FNA and MonoGame. I defaulted the project to MonoGame, since MonoGame is available as a NuGet package and only requires 1 native library (SDL2). Updated README.md with steps to take to get it working with FNA and renamed the project itself to ImGui.NET.SampleProgram.XNA, since it's no longer FNA-only.

Although a couple of extra buffers were necessary to copy the draw data, I'm caching them and resizing when needed, along with the gpu buffers. The copy itself is done using the method you suggested.

Interesting sidenote: when the project is compiled against MonoGame, a deprecation warning is spawned telling us that the overload with the "minVertexIndex"- and "numVertices"-arguments is obsolete and will be removed in the future.

![image](https://user-images.githubusercontent.com/1295673/44308887-339f6d00-a3be-11e8-95d4-965dcc7d0d6e.png)

The version without these arguments is not available in FNA, so I opted to keep this method nonetheless. But it confirms the discussion we had earlier.

mellinoe commented:
The latest version is good; it's helpful to be able to grab MonoGame with NuGet.

Can we suppress the warning emitted?

```C#
#pragma warning disable 0618 // FNA does not expose an alternative method.
...
#pragma warning restore
```

The rest of it looks good and seems to run correctly. Thanks for all the work on this!

flyingpie commented:
@mellinoe Done! Thank you for your suggestions, I think we made a nice example renderer here.

mellinoe commented:
@flyingpie Awesome work. Now I have something to point folks towards when they ask about MonoGame/FNA support.
